Firewall creation failure during initial attempt

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

<!--- Please keep this note for the community --->

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

<!--- Thank you for keeping this note for the community --->


### Terraform Version

v1.3.7

### AzureRM Provider Version

3.69.0

### Affected Resource(s)/Data Source(s)

azurerm_palo_alto_next_generation_firewall_virtual_network_local_rulestack

### Terraform Configuration Files

```hcl
# create resource group
resource azurerm_resource_group "rg" {
    name = "rg-${var.resource_suffix}"
    location = var.location
    tags = var.tags
} 

# create vnet/subnets

resource "azurerm_virtual_network" "vnet" {
  name                = "vnet-fw-${var.resource_suffix}"
  address_space       = [var.vnet_cidr]
  location            = var.location

  resource_group_name = azurerm_resource_group.rg.name

  tags = {
    Purpose = "terraform testing"
  }
}

# create localrulestack
resource azurerm_palo_alto_local_rulestack "lrs1" {
    name = "lrs-${var.resource_suffix}-1"
    resource_group_name = azurerm_resource_group.rg.name
    location = var.location
    description = "rulestack1"
}

resource azurerm_palo_alto_local_rulestack "lrs2" {
    name = "lrs-${var.resource_suffix}-2"
    resource_group_name = azurerm_resource_group.rg.name
    location = var.location
    description = "rulestack2"
}

# create prefix list
resource "azurerm_palo_alto_local_rulestack_prefix_list" "prefix" {
  name         = "prefix-${var.resource_suffix}"
  prefix_list  = [ "10.0.2.0/26" ]
  rulestack_id = azurerm_palo_alto_local_rulestack.lrs1.id
  description = "prefixlist1"

  depends_on = [ azurerm_palo_alto_local_rulestack.lrs1 ]
}

# create fqdn list
resource azurerm_palo_alto_local_rulestack_fqdn_list "fqdn" {
  name = "fqdn-${var.resource_suffix}"
  rulestack_id = azurerm_palo_alto_local_rulestack.lrs1.id
  fully_qualified_domain_names = ["www.defnews.com"]
  depends_on = [ azurerm_palo_alto_local_rulestack.lrs1 ]
}

# create certificate (self signed)
resource "azurerm_palo_alto_local_rulestack_certificate" "cert" {
  name         = "cert-${var.resource_suffix}"
  rulestack_id = azurerm_palo_alto_local_rulestack.lrs1.id
  self_signed = true
  # key_vault_certificate_id  = "https://kvtesting456.vault.azure.net/certificates/kvtesting"
  depends_on = [ azurerm_palo_alto_local_rulestack.lrs1 ]
}

# creare local rulestack rule
resource "azurerm_palo_alto_local_rulestack_rule" "localrule" {
  name          = "rule-${var.resource_suffix}"
  rulestack_id = azurerm_palo_alto_local_rulestack.lrs1.id
  priority      = 1000
  
  applications  = ["any"]

  source {
    cidrs = ["192.168.16.0/24"]
  }
  destination {
    cidrs = ["192.168.16.0/24"]
  }
  action = "Allow"

  depends_on = [ azurerm_palo_alto_local_rulestack.lrs1 ]
  tags = var.tags
}

# create localrulestack rule2
resource "azurerm_palo_alto_local_rulestack_rule" "localrule2" {
  name          = "rule2-${var.resource_suffix}"
  rulestack_id = azurerm_palo_alto_local_rulestack.lrs1.id
  priority      = 1001
  
  applications  = ["any"]

  source {
    cidrs = ["192.168.16.0/25"]
  }
  destination {
    cidrs = ["192.168.16.0/25"]
  }
  action = "Allow"

  depends_on = [ azurerm_palo_alto_local_rulestack.lrs1 ]
  tags = var.tags
}

resource "azurerm_palo_alto_local_rulestack_rule" "localrule3" {
  name          = "rule3-${var.resource_suffix}"
  rulestack_id = azurerm_palo_alto_local_rulestack.lrs1.id
  priority      = 1002
  
  applications  = ["any"]

  source {
    cidrs = ["192.168.16.0/25"]
  }
  destination {
    cidrs = ["192.168.16.0/25"]
  }
  action = "Allow"

  depends_on = [ azurerm_palo_alto_local_rulestack.lrs1 ]
  tags = var.tags
}

resource "azurerm_palo_alto_local_rulestack_rule" "localrule5" {
  name          = "rule5-${var.resource_suffix}"
  rulestack_id = azurerm_palo_alto_local_rulestack.lrs1.id
  priority      = 1003
  
  applications  = ["any"]

  source {
    cidrs = ["192.168.16.0/25"]
  }
  destination {
    cidrs = ["192.168.16.0/25"]
  }
  action = "Allow"

  depends_on = [ azurerm_palo_alto_local_rulestack.lrs1 ]
  tags = var.tags
}

resource "azurerm_palo_alto_local_rulestack_rule" "localrule4" {
  name          = "rule4-${var.resource_suffix}"
  rulestack_id = azurerm_palo_alto_local_rulestack.lrs1.id
  priority      = 1004
  
  applications  = ["any"]

  source {
    cidrs = ["192.168.16.0/25"]
  }
  destination {
    cidrs = ["192.168.16.0/25"]
  }
  action = "Allow"

  depends_on = [ azurerm_palo_alto_local_rulestack.lrs1 ]
  tags = var.tags
}

data "azurerm_virtual_network" "vnet" {
  name = "vnet-fw-${var.resource_suffix}"
  resource_group_name = var.rgname
}

data "azurerm_palo_alto_local_rulestack" "lrs-out" {
  resource_group_name = var.rgname
  name = var.lrsname
}

resource "azurerm_subnet" "trust" {
  name                 = "subnet-trust-vnetfw-${var.resource_suffix}"
  resource_group_name  = var.rgname
  virtual_network_name = var.vnet_name
  address_prefixes     = [var.trust_subnet_cidr]

  delegation {
    name = "trusted"

    service_delegation {
      name = "PaloAltoNetworks.Cloudngfw/firewalls"
      actions = [
        "Microsoft.Network/virtualNetworks/subnets/join/action",
      ]
    }
  }
}

resource "azurerm_subnet" "untrust" {
  name                 = "subnet-untrust-vnetfw-${var.resource_suffix}"
  resource_group_name  = var.rgname
  virtual_network_name = var.vnet_name
  address_prefixes     = [var.untrust_subnet_cidr]

  delegation {
    name = "untrusted"

    service_delegation {
      name = "PaloAltoNetworks.Cloudngfw/firewalls"
      actions = [
        "Microsoft.Network/virtualNetworks/subnets/join/action",
      ]
    }
  }
}

# create outbound certificate association
resource "azurerm_palo_alto_local_rulestack_outbound_trust_certificate_association" "certassoc" {
  certificate_id = var.cert_id
}

# create public ip
resource "azurerm_public_ip" "vnetfw-pubip" {
  name                ="pip-vnetfw-${var.resource_suffix}"
  location            = var.location
  resource_group_name = var.rgname
  allocation_method   = "Static"
  sku                 = "Standard"
}

# create vnet firewall
resource "azurerm_palo_alto_next_generation_firewall_virtual_network_local_rulestack" "firewall1" {
  name                = "vnetfw-${var.resource_suffix}"
  resource_group_name = var.rgname
  rulestack_id        = data.azurerm_palo_alto_local_rulestack.lrs-out.id

  network_profile {
    public_ip_address_ids = [azurerm_public_ip.vnetfw-pubip.id]

    vnet_configuration {
      virtual_network_id  = data.azurerm_virtual_network.vnet.id
      trusted_subnet_id   = azurerm_subnet.trust.id
      untrusted_subnet_id = azurerm_subnet.untrust.id
    }
  }

  dns_settings {
    dns_servers = [azurerm_public_ip.vnetfw-pubip.ip_address]
  }

  # destination_nat {
  #   backend_config {
  #     port = 80
  #     public_ip_address = "10.3.1.12"
  #   }
  #   frontend_config {
  #     port = 80
  #     public_ip_address_id = azurerm_public_ip.fw-vnet-pubip.id
  #   }
  #   protocol = "TCP"
  #   name = "${var.resource_prefix}-fw-${local.timestring}-dnat1"
  # }
  tags = var.tags
}
```


### Debug Output/Panic Output

```shell
“tenantId”: “e6579873-b3c7-4248-b0b5-700db290acec”,
    “properties”: {
        “statusCode”: “BadRequest”,
        “serviceRequestId”: null,
        “statusMessage”: “{\“error\“:{\“code\“:\“PANError\“,\“message\“:\“Commit process unfinished.\“}}“,
        “eventCategory”: “Administrative”,
```


### Expected Behaviour

Able to create firewall along with rulestack. Firewall creation should proceed after successfull commit operation performed on rulestack,.

### Actual Behaviour

Create all the resources (rulestack, rules. fqdn, prefix , firewall) from a single tf file -> this gives the above error.

Instead if we create rulestack related resources first and then firewall creation -> this works.

Upon investigation, we noticed that firewall creation is initiated when rulestack commit was still not complete triggerring race conditions.

### Steps to Reproduce

1. Create firewall and all other resources (rulestack,rules etc..) from a single tf file.

### Important Factoids

_No response_

### References

_No response_

The real cause of the issue should be the process of certificate commit failure. This may caused by a server-side error, or a certificate configuration mistake.


```
2023-10-10T11:18:31.785-0700 [ERROR] vertex "azurerm_palo_alto_local_rulestack_certificate.cert" error: committing Local RuleStack config for Local Rulestack Certificate (Subscription: "xxx"
Resource Group Name: "rg-tfTest02"
Local Rulestack Name: "lrs-tfTest02-1"
Certificate Name: "cert-tfTest02"): polling after Commit: polling failed: the Azure API returned the following error:

Status: "Failed"
Code: "RequestTimeout"
Message: "The operation did not complete in maximum retry"
Activity Id: ""

---

API Response:

----[start]----
{"id":"/providers/PaloAltoNetworks.Cloudngfw/locations/NORTHCENTRALUS/operationStatuses/xx","name":"xxx","resourceId":"/subscriptions/xxx/resourceGroups/rg-tfTest02/providers/PaloAltoNetworks.Cloudngfw/localRulestacks/lrs-tfTest02-1","status":"Failed","startTime":"2023-10-10T18:12:55.5548446Z","endTime":"2023-10-10T18:18:30.2242916Z","error":{"code":"RequestTimeout","message":"The operation did not complete in maximum retry"}}
-----[end]-----

```

--- a/internal/services/paloalto/next_generation_firewall_vhub_local_rulestack_resource.go
+++ b/internal/services/paloalto/next_generation_firewall_vhub_local_rulestack_resource.go
@@ -80,7 +80,7 @@ func (r NextGenerationFirewallVHubLocalRuleStackResource) Attributes() map[strin
 
 func (r NextGenerationFirewallVHubLocalRuleStackResource) Create() sdk.ResourceFunc {
 	return sdk.ResourceFunc{
-		Timeout: 2 * time.Hour,
+		Timeout: 3 * time.Hour,
 		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
 			client := metadata.Client.PaloAlto.Client.Firewalls
 			localRuleStackClient := metadata.Client.PaloAlto.Client.LocalRulestacks
@@ -222,7 +222,7 @@ func (r NextGenerationFirewallVHubLocalRuleStackResource) Delete() sdk.ResourceF
 
 func (r NextGenerationFirewallVHubLocalRuleStackResource) Update() sdk.ResourceFunc {
 	return sdk.ResourceFunc{
-		Timeout: 2 * time.Hour,
+		Timeout: 3 * time.Hour,
 		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
 			client := metadata.Client.PaloAlto.Client.Firewalls
 
@@ -278,6 +278,12 @@ func (r NextGenerationFirewallVHubLocalRuleStackResource) Update() sdk.ResourceF
 
 			if metadata.ResourceData.HasChange("tags") {
 				firewall.Tags = tags.Expand(model.Tags)
+			}
+
+			if metadata.ResourceData.HasChange("rulestack_id") {
+				ruleStackID, _ := localrulestacks.ParseLocalRulestackID(model.RuleStackId)
+				locks.ByID(ruleStackID.ID())
+				defer locks.UnlockByID(ruleStackID.ID())
 			}
 
 			if err = client.CreateOrUpdateThenPoll(ctx, *id, firewall); err != nil {

--- a/internal/services/paloalto/next_generation_firewall_vhub_local_rulestack_resource.go
+++ b/internal/services/paloalto/next_generation_firewall_vhub_local_rulestack_resource.go
@@ -281,7 +281,10 @@ func (r NextGenerationFirewallVHubLocalRuleStackResource) Update() sdk.ResourceF
 			}
 
 			if metadata.ResourceData.HasChange("rulestack_id") {
-				ruleStackID, _ := localrulestacks.ParseLocalRulestackID(model.RuleStackId)
+				ruleStackID, err := localrulestacks.ParseLocalRulestackID(model.RuleStackId)
+				if err != nil {
+					return fmt.Errorf("parsing rulestack_id %s: %+v", model.RuleStackId, err)
+				}
 				locks.ByID(ruleStackID.ID())
 				defer locks.UnlockByID(ruleStackID.ID())
 			}

--- a/internal/services/paloalto/next_generation_firewall_vhub_local_rulestack_resource.go
+++ b/internal/services/paloalto/next_generation_firewall_vhub_local_rulestack_resource.go
@@ -12,6 +12,7 @@ import (
 	"github.com/hashicorp/go-azure-helpers/resourcemanager/tags"
 	"github.com/hashicorp/go-azure-sdk/resource-manager/paloaltonetworks/2022-08-29/firewalls"
 	"github.com/hashicorp/go-azure-sdk/resource-manager/paloaltonetworks/2022-08-29/localrulestacks"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/locks"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/services/paloalto/schema"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/services/paloalto/validate"
@@ -136,6 +137,9 @@ func (r NextGenerationFirewallVHubLocalRuleStackResource) Create() sdk.ResourceF
 
 				Tags: tags.Expand(model.Tags),
 			}
+
+			locks.ByID(ruleStackID.ID())
+			defer locks.UnlockByID(ruleStackID.ID())
 
 			if err = client.CreateOrUpdateThenPoll(ctx, id, firewall); err != nil {
 				return err

--- a/internal/services/paloalto/next_generation_firewall_vhub_local_rulestack_resource.go
+++ b/internal/services/paloalto/next_generation_firewall_vhub_local_rulestack_resource.go
@@ -260,6 +260,8 @@ func (r NextGenerationFirewallVHubLocalRuleStackResource) Update() sdk.ResourceF
 				}
 
 				props.AssociatedRulestack = ruleStack
+				locks.ByID(ruleStackID.ID())
+				defer locks.UnlockByID(ruleStackID.ID())
 			}
 
 			if metadata.ResourceData.HasChange("network_profile") {
@@ -278,15 +280,6 @@ func (r NextGenerationFirewallVHubLocalRuleStackResource) Update() sdk.ResourceF
 
 			if metadata.ResourceData.HasChange("tags") {
 				firewall.Tags = tags.Expand(model.Tags)
-			}
-
-			if metadata.ResourceData.HasChange("rulestack_id") {
-				ruleStackID, err := localrulestacks.ParseLocalRulestackID(model.RuleStackId)
-				if err != nil {
-					return fmt.Errorf("parsing rulestack_id %s: %+v", model.RuleStackId, err)
-				}
-				locks.ByID(ruleStackID.ID())
-				defer locks.UnlockByID(ruleStackID.ID())
 			}
 
 			if err = client.CreateOrUpdateThenPoll(ctx, *id, firewall); err != nil {

--- a/internal/services/paloalto/next_generation_firewall_vnet_local_rulestack_resource.go
+++ b/internal/services/paloalto/next_generation_firewall_vnet_local_rulestack_resource.go
@@ -12,6 +12,7 @@ import (
 	"github.com/hashicorp/go-azure-helpers/resourcemanager/tags"
 	"github.com/hashicorp/go-azure-sdk/resource-manager/paloaltonetworks/2022-08-29/firewalls"
 	"github.com/hashicorp/go-azure-sdk/resource-manager/paloaltonetworks/2022-08-29/localrulestacks"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/locks"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/services/paloalto/schema"
 	"github.com/hashicorp/terraform-provider-azurerm/internal/services/paloalto/validate"
@@ -130,6 +131,9 @@ func (r NextGenerationFirewallVNetLocalRulestackResource) Create() sdk.ResourceF
 				},
 				Tags: tags.Expand(model.Tags),
 			}
+
+			locks.ByID(ruleStackID.ID())
+			defer locks.UnlockByID(ruleStackID.ID())
 
 			if err = client.CreateOrUpdateThenPoll(ctx, id, firewall); err != nil {
 				return fmt.Errorf("creating %s: %+v", id, err)

--- a/internal/services/paloalto/next_generation_firewall_vnet_local_rulestack_resource.go
+++ b/internal/services/paloalto/next_generation_firewall_vnet_local_rulestack_resource.go
@@ -276,7 +276,10 @@ func (r NextGenerationFirewallVNetLocalRulestackResource) Update() sdk.ResourceF
 			}
 
 			if metadata.ResourceData.HasChange("rulestack_id") {
-				ruleStackID, _ := localrulestacks.ParseLocalRulestackID(model.RuleStackId)
+				ruleStackID, err := localrulestacks.ParseLocalRulestackID(model.RuleStackId)
+				if err != nil {
+					return fmt.Errorf("parsing rulestack_id %s: %+v", model.RuleStackId, err)
+				}
 				locks.ByID(ruleStackID.ID())
 				defer locks.UnlockByID(ruleStackID.ID())
 			}

--- a/internal/services/paloalto/next_generation_firewall_vnet_local_rulestack_resource.go
+++ b/internal/services/paloalto/next_generation_firewall_vnet_local_rulestack_resource.go
@@ -255,6 +255,8 @@ func (r NextGenerationFirewallVNetLocalRulestackResource) Update() sdk.ResourceF
 				}
 
 				props.AssociatedRulestack = ruleStack
+				locks.ByID(ruleStackID.ID())
+				defer locks.UnlockByID(ruleStackID.ID())
 			}
 
 			if metadata.ResourceData.HasChange("network_profile") {
@@ -273,15 +275,6 @@ func (r NextGenerationFirewallVNetLocalRulestackResource) Update() sdk.ResourceF
 
 			if metadata.ResourceData.HasChange("tags") {
 				firewall.Tags = tags.Expand(model.Tags)
-			}
-
-			if metadata.ResourceData.HasChange("rulestack_id") {
-				ruleStackID, err := localrulestacks.ParseLocalRulestackID(model.RuleStackId)
-				if err != nil {
-					return fmt.Errorf("parsing rulestack_id %s: %+v", model.RuleStackId, err)
-				}
-				locks.ByID(ruleStackID.ID())
-				defer locks.UnlockByID(ruleStackID.ID())
 			}
 
 			if err = client.CreateOrUpdateThenPoll(ctx, *id, firewall); err != nil {

--- a/internal/services/paloalto/next_generation_firewall_vnet_local_rulestack_resource.go
+++ b/internal/services/paloalto/next_generation_firewall_vnet_local_rulestack_resource.go
@@ -75,7 +75,7 @@ func (r NextGenerationFirewallVNetLocalRulestackResource) ResourceType() string 
 
 func (r NextGenerationFirewallVNetLocalRulestackResource) Create() sdk.ResourceFunc {
 	return sdk.ResourceFunc{
-		Timeout: 2 * time.Hour,
+		Timeout: 3 * time.Hour,
 		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
 			client := metadata.Client.PaloAlto.Client.Firewalls
 			localRulestackClient := metadata.Client.PaloAlto.Client.LocalRulestacks
@@ -216,7 +216,7 @@ func (r NextGenerationFirewallVNetLocalRulestackResource) IDValidationFunc() plu
 
 func (r NextGenerationFirewallVNetLocalRulestackResource) Update() sdk.ResourceFunc {
 	return sdk.ResourceFunc{
-		Timeout: 2 * time.Hour,
+		Timeout: 3 * time.Hour,
 		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
 			client := metadata.Client.PaloAlto.Client.Firewalls
 
@@ -273,6 +273,12 @@ func (r NextGenerationFirewallVNetLocalRulestackResource) Update() sdk.ResourceF
 
 			if metadata.ResourceData.HasChange("tags") {
 				firewall.Tags = tags.Expand(model.Tags)
+			}
+
+			if metadata.ResourceData.HasChange("rulestack_id") {
+				ruleStackID, _ := localrulestacks.ParseLocalRulestackID(model.RuleStackId)
+				locks.ByID(ruleStackID.ID())
+				defer locks.UnlockByID(ruleStackID.ID())
 			}
 
 			if err = client.CreateOrUpdateThenPoll(ctx, *id, firewall); err != nil {

--- a/website/docs/r/palo_alto_next_generation_firewall_vhub_local_rulestack.html.markdown
+++ b/website/docs/r/palo_alto_next_generation_firewall_vhub_local_rulestack.html.markdown
@@ -148,9 +148,9 @@ In addition to the Arguments listed above - the following Attributes are exporte
 
 The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/language/resources/syntax#operation-timeouts) for certain actions:
 
-* `create` - (Defaults to 2 hours) Used when creating the Palo Alto Next Generation Firewall VHub Local Rulestack.
+* `create` - (Defaults to 3 hours) Used when creating the Palo Alto Next Generation Firewall VHub Local Rulestack.
 * `read` - (Defaults to 5 minutes) Used when retrieving the Palo Alto Next Generation Firewall VHub Local Rulestack.
-* `update` - (Defaults to 2 hours) Used when updating the Palo Alto Next Generation Firewall VHub Local Rulestack.
+* `update` - (Defaults to 3 hours) Used when updating the Palo Alto Next Generation Firewall VHub Local Rulestack.
 * `delete` - (Defaults to 2 hours) Used when deleting the Palo Alto Next Generation Firewall VHub Local Rulestack.
 
 ## Import

--- a/website/docs/r/palo_alto_next_generation_firewall_virtual_network_local_rulestack.html.markdown
+++ b/website/docs/r/palo_alto_next_generation_firewall_virtual_network_local_rulestack.html.markdown
@@ -217,9 +217,9 @@ In addition to the Arguments listed above - the following Attributes are exporte
 
 The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/language/resources/syntax#operation-timeouts) for certain actions:
 
-* `create` - (Defaults to 2 hours) Used when creating the Palo Alto Next Generation Firewall Virtual Network Local Rulestack.
+* `create` - (Defaults to 3 hours) Used when creating the Palo Alto Next Generation Firewall Virtual Network Local Rulestack.
 * `read` - (Defaults to 5 minutes) Used when retrieving the Palo Alto Next Generation Firewall Virtual Network Local Rulestack.
-* `update` - (Defaults to 2 hours) Used when updating the Palo Alto Next Generation Firewall Virtual Network Local Rulestack.
+* `update` - (Defaults to 3 hours) Used when updating the Palo Alto Next Generation Firewall Virtual Network Local Rulestack.
 * `delete` - (Defaults to 2 hours) Used when deleting the Palo Alto Next Generation Firewall Virtual Network Local Rulestack.
 
 ## Import

